WordPress/gutenberg
# Core Data: specify context=edit only when needed

## Description

API requests for fetching entities currently add `context=edit` to all requests. Some endpoints do not use this parameter, which can cause unexpected problems, like the need to preload both the `/` and `/?context=edit` endpoints (which both have exactly the same response).

This change allows specifying query parameters specific to each entity, and adds `context=edit` only to the entities that use that parameter.

This resolves the (rejected) Core ticket (https://core.trac.wordpress.org/ticket/50606) and removes the associated filter for preloading `/?context=edit`.

Follows up on https://github.com/WordPress/gutenberg/pull/22952.

## How has this been tested?

1. Notice that the site icon will load in the editor without prefetching the `/?context=edit` endpoint.
1. Existing resolver unit tests pass with minimal changes.
1. Verify that the `context=edit` parameter is added [only to the endpoints](https://developer.wordpress.org/rest-api/reference/) where it's a valid parameter
1. Verify that the [entity urls generated by the resolver](https://github.com/WordPress/gutenberg/pull/30482/files#diff-fec6390e24332dec4a0c7c74f9f7b9a015ce191e0e6348361ead793ceda3991dR113) match the [editor preload paths](https://github.com/WordPress/WordPress/blob/cc0ce2c91d9b59c131012f1d8d648705486bc42b/wp-admin/edit-form-blocks.php#L46).
1. Verify that editor loads and works as expected (no failed API calls or missing data that should be preloaded)

You can also compare [resolver paths for entities](https://github.com/WordPress/gutenberg/pull/30482/files#diff-fec6390e24332dec4a0c7c74f9f7b9a015ce191e0e6348361ead793ceda3991dR113) with and without this branch applied, for example:

| Kind | Name | Key | Path (trunk) | Path (this branch) |
| - | - | - | - | - |
| root | __unstableBase | '' | `/?context=edit` | `/` |
| root | site | '' | `/wp/v2/settings/?context=edit` | `/wp/v2/settings` |
| root | postType | post | `/wp/v2/types/post?context=edit` | `/wp/v2/types/post?context=edit` |
| postType | wp_template | tt1-blocks//single | `/wp/v2/templates/tt1-blocks//single?context=edit` | `/wp/v2/templates/tt1-blocks//single?context=edit` |
| postType | wp_global_styles | 26 | `/wp/v2/__experimental/global-styles/26?context=edit` | `/wp/v2/__experimental/global-styles/26?context=edit` |
| root | media | 10 | `/wp/v2/media/10?context=edit` | `/wp/v2/media/10?context=edit` |
| root | taxonomy | category | `/wp/v2/taxonomies/category?context=edit` | `/wp/v2/taxonomies/category?context=edit` |
| root | taxonomy | post_tag | `/wp/v2/taxonomies/post_tag?context=edit` | `/wp/v2/taxonomies/post_tag?context=edit` |
| root | user | 1 | `/wp/v2/users/1?context=edit` | `/wp/v2/users/1?context=edit` |

Note how `/` and `/settings` correctly no longer use `?context=edit`, but other endpoints still do.

## Screenshots 

Site icon displayed in upper left corner:

<img width="681" alt="image" src="https://user-images.githubusercontent.com/1699996/113465370-1433a200-93f9-11eb-98e5-c254afa7ca51.png">


## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
